core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,397 @@
+   * None essentially means -Infinity as either there are no linked expressions
+   * or there are no lambda variables used in the linked expressions
+   */
+  def maxLambdaVar: Option[Int]
+}
+
+object NormalExpression {
+  case class App(fn: NormalExpression, arg: NormalExpression)
+  extends NormalExpression {
+    val maxLambdaVar = (fn.maxLambdaVar.toList ++ arg.maxLambdaVar.toList)
+      .reduceLeftOption(Math.max)
+  }
+  case class ExternalVar(pack: PackageName, defName: Identifier)
+  extends NormalExpression {
+    val maxLambdaVar = None
+  }
+  case class Match(arg: NormalExpression,
+    branches: NonEmptyList[(Pattern[(PackageName, Constructor), Type], NormalExpression)])
+  extends NormalExpression {
+    val maxLambdaVar =
+      (arg.maxLambdaVar.toList ++ branches.toList.flatMap(_._2.maxLambdaVar))
+        .reduceLeftOption(Math.max)
+  }
+  case class LambdaVar(index: Int) extends NormalExpression {
+    val maxLambdaVar = Some(index)
+  }
+  /*
+   * It is reasonable to ask how you can define a lambda without an identifier
+   * for its argument in its expression. This is a benefit of de bruijn indexing.
+   * When all lambdas have exactly one argument you identify the var by how many
+   * lambdas out you have to travel.
+   *
+   * eg \x -> \y -> [y,x] would have normalization Lambda(Lambda(Apply(Apply(List, LambdaVar(1)), LambdaVar(0))))
+   *
+   * ref: https://en.wikipedia.org/wiki/De_Bruijn_index
+   */
+  case class Lambda(expr: NormalExpression) extends NormalExpression {
+    val maxLambdaVar = expr.maxLambdaVar.map(_ - 1)
+  }
+  case class Struct(enum: Int, args: List[NormalExpression]) extends NormalExpression {
+    val maxLambdaVar = args.flatMap(_.maxLambdaVar).reduceLeftOption(Math.max)
+  }
+  case class Literal(lit: Lit) extends NormalExpression {
+    val maxLambdaVar = None
+  }
+  case class Recursion(lambda: NormalExpression) extends NormalExpression {
+    val maxLambdaVar = lambda.maxLambdaVar
+  }
+}
+
+object Normalization {
+  case class NormalExpressionTag(ne: NormalExpression, children: Set[NormalExpression])
+  type NormalizedPM = PackageMap.Typed[(Declaration, Normalization.NormalExpressionTag)]
+  type NormalizedPac = Package.Typed[(Declaration, Normalization.NormalExpressionTag)]
+
+  private def normalOrderReduction(expr: NormalExpression): NormalExpression = {
+    import NormalExpression._
+    val nextExpr = expr match {
+      case App(Lambda(nextExpr), arg) => {
+        applyLambdaSubstituion(nextExpr, Some(arg), 0)
+      }
+      case App(fn, arg) => {
+        val nextFn = normalOrderReduction(fn)
+        App(nextFn, arg)
+      }
+      /*case Match(struct@Struct(enum, args), branches) => { TODO: If s matches a pattern we should perform lambdaSubstitution
+        Match(struct, branches)
+      }*/
+      case Recursion(Lambda(innerExpr)) if(innerExpr.maxLambdaVar.map(_ < 0).getOrElse(false)) => {
```

Comment:
nit: space after if, but actually no parens needed in case matches:

```scala
if innerExpr.maxLambdaVar.exists(_ < 0) =>
```
what is a negative index? Is it special? Can we remind the reader here with a comment?

Reply:
This is a great question and i will add a comment. You can't have a negative index (and I thiiiiink I could prove that with the type system. Just need to give some care in `lambdaSubstitution` where there is a decrement... but I think that can be shown to be non-negative. But `maxLambdaVar` can. That's because the `maxLambdaVar` of a `Lambda` is the decrement of its expression. If an expression has a negative `maxLambdaVar` then all the variables are bounded internally and wrapping with an `App(Lambda` will have no effect (it'll just thunk).

In the `App(Lambda` case it's captured with `beta` reduction. But I've got this `Recursion` expression for a solution to recursion.